kube-client/src/api/entry.rs

```diff
@@ -194,13 +201,89 @@ impl<'a, K> OccupiedEntry<'a, K> {
         Ok(())
     }
+
+    /// Validate that [`Self::object`] is valid, and refers to the same object as the original [`Api::entry`] call
+    ///
+    /// Defaults [`ObjectMeta::name`] and [`ObjectMeta::namespace`] if unset.
+    fn prepare_for_commit(&mut self) -> Result<(), CommitValidationError>
+    where
+        K: Resource,
+    {
+        // Access `Self::object` directly rather than using `Self::get_mut` to avoid flagging the object as dirty
+        let meta = self.object.meta_mut();
+        match &mut meta.name {
+            name @ None => *name = Some(self.name.to_string()),
+            Some(name) if name != self.name => {
+                return Err(CommitValidationError::NameMismatch {
+                    object_name: name.clone(),
+                    expected: self.name.to_string(),
+                });
+            }
+            Some(_) => (),
+        }
+        match &mut meta.namespace {
+            ns @ None => *ns = self.api.namespace.clone(),
+            Some(ns) if Some(ns.as_str()) != self.api.namespace.as_deref() => {
+                return Err(CommitValidationError::NamespaceMismatch {
+                    object_namespace: Some(ns.clone()),
+                    expected: self.api.namespace.clone(),
+                });
+            }
+            Some(_) => (),
+        }
+        if let Some(generate_name) = &meta.generate_name {
+            return Err(CommitValidationError::GenerateName {
+                object_generate_name: generate_name.clone(),
+            });
+        }
+        Ok(())
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+/// Commit errors
+pub enum CommitError {
+    /// Pre-commit validation failed
+    #[error("failed to validate object for saving")]
+    Validate(#[from] CommitValidationError),
+    /// Failed to submit the new object to the Kubernetes API
+    #[error("failed to save object")]
+    Save(#[source] Error),
+}
+
+#[derive(Debug, thiserror::Error)]
+/// Pre-commit validation errors
+pub enum CommitValidationError {
+    /// [`ObjectMeta::name`] does not match the name passed to [`Api::entry`]
+    #[error(".metadata.name does not match the name passed to Api::entry (got: {object_name:?}, expected: {expected:?})")]
+    NameMismatch {
+        /// The name of the object ([`ObjectMeta::name`])
+        object_name: String,
+        /// The name passed to [`Api::entry`]
+        expected: String,
+    },
+    /// [`ObjectMeta::namespace`] does not match the namespace of the [`Api`]
+    #[error(".metadata.namespace does not match the namespace of the Api (got: {object_namespace:?}, expected: {expected:?})")]
+    NamespaceMismatch {
+        /// The name of the object ([`ObjectMeta::namespace`])
+        object_namespace: Option<String>,
+        /// The namespace of the [`Api`]
+        expected: Option<String>,
+    },
+    /// [`ObjectMeta::generate_name`] must not be set
+    #[error(".metadata.generate_name must not be set (got: {object_generate_name:?})")]
```

Comment:
these errors look great. however, they do use the new 1.58 f-strings so will need to do the msrv bump: https://kube-rs.github.io/website/rust-version/#minimum-supported-rust-version

Reply:
`thiserror` has had specific support for this syntax since before Rust mainlined it. For example, I still use rustc 1.56.1 as my primary local compiler (due to #746), and it builds fine there.